### PR TITLE
ARM Template to launch Cohesive Networks VNS3 network security appliance demo environment

### PR DESCRIPTION
**Pull request by Cohesive-DevTeam**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: cohesive-vns3-free-multiclient-overlay-linux/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Fcohesive-vns3-free-multiclient-overlay-linux%2Fazuredeploy.json" target="_blank">
+<img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.png"/>
+</a>
+<a href="http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Fcohesive-vns3-free-multiclient-overlay-linux%2Fazuredeploy.json" target="_blank">
+<img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.png"/>
+</a>
+
+`Tags: VPN, HA, Multicast, Network Traffic Analysis, Network Visibility, Federation, Security, Isolation, Segmentation, Firewall, NIDS, WAF, Load-balancer, multi-cloud, hybrid, UTM, overlay, awesome`
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/README.md]: move tags to the bottom of the readme

File: cohesive-vns3-free-multiclient-overlay-linux/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+# VNS3 cloud networking appliance for security, connectivity and federation in the clouds
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/README.md]: move header to the top of the file and add badges

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+        "virtualNetworkName": "[concat(variables('resourceGroupName'), '-vnet')]",
+        "vns3ControllerName": "VNS3Controller",
+        "networkInterfaceName": "[toLower(concat(variables('vns3ControllerName'), '-nic'))]",
+        "publicIPAddresseName": "[concat(variables('vns3ControllerName'), '-ip')]",
+        "networkSecurityGroupName": "[concat(variables('vns3ControllerName'), '-nsg')]",
+        "virtualSubnetName": "VNS3_VirtualSubnet",
+        "vnetSubnet":"10.10.10.0/28",
+        "vnsStaticIp": "10.10.10.10",
+        "CustomScriptForLinux": "CustomScript",
+        "artifactsLocation":"https://raw.githubusercontent.com/Cohesive-DevTeam/Quickstart/master/",
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+            "location": "[parameters('location')]",
+            "plan": {
+                "name": "cohesive-vns3-4_4_x-free",
+                "product": "vns3_4x_network_security",
+                "publisher": "cohesive"
+            },
+            "tags": {
+                "Name": "VNS3 Controller"
+            },
+            "scale": null,
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: remove all empty/null properties

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+                    "networkInterfaces": [
+                        {
+                            "id": "[resourceId('Microsoft.Network/networkInterfaces', variables('networkInterfaceName'))]"
+                        }
+                    ]
+                },
+                "diagnosticsProfile": {
+                    "bootDiagnostics": {
+                        "enabled": true,
+                        "storageUri": "[reference(variables('storageAccount'),'2017-10-01').primaryEndpoints['blob']]"
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: update apiVersion

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+                    }
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: move this above properties

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+        },
+        {
+            "type": "Microsoft.Network/networkInterfaces",
+            "name": "[variables('networkInterfaceName')]",
+            "apiVersion": "2019-04-01",
+            "location": "[parameters('location')]",
+            "tags": {
+                "Name": "VNS3 Controller"
+            },
+            "scale": null,
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: remove all empty/null properties - won't comment on the rest

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+                        "properties": {
+                            "privateIPAllocationMethod": "Static",
+                            "privateIPAddress": "[variables('vnsStaticIp')]",
+                            "publicIPAddress": {
+                                "id": "[resourceId('Microsoft.Network/publicIPAddresses', variables('publicIPAddresseName'))]"
+                            },
+                            "subnet": {
+                                "id": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('virtualNetworkName'), variables('virtualSubnetName'))]"
+                            },
+                            "primary": true,
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: remove this

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+                            "privateIPAddressVersion": "IPv4"
+                        }
+                    }
+                ],
+                "enableAcceleratedNetworking": false,
+                "enableIPForwarding": true,
+                "networkSecurityGroup": {
+                    "id": "[resourceId('Microsoft.Network/networkSecurityGroups', variables('networkSecurityGroupName'))]"
+                },
+                "primary": true
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: remove this

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: move all dependsOn above properties on all the resources (won't comment on the rest)

File: cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+                            "priority": 400,
+                            "direction": "Inbound"
+                        }
+                    }
+                ]
+            }
+        },
+        {
+            "type": "Microsoft.Network/publicIPAddresses",
+            "sku": {
```
Comment (bmoore-msft) [inline cohesive-vns3-free-multiclient-overlay-linux/azuredeploy.json]: move this below location - do the same on the other resources below

**Comment by HKF1977** [comment]
@bmoore-msft Thanks for reviewing, I've made the changes as requested. + reordered inline with new contribution guide. But I'm struggling with the new errors, any pointers would be handy. Thanks Scott

**Pull request merged**